File: src/cpp/src/formats/vtm.cpp
//  ██████   ██████ ██████████  █████████  █████   █████ █████    ███████
// ░░██████ ██████ ░░███░░░░░█ ███░░░░░███░░███   ░░███ ░░███   ███░░░░░███      ███         ███
//  ░███░█████░███  ░███  █ ░ ░███    ░░░  ░███    ░███  ░███  ███     ░░███    ░███        ░███
//  ░███░░███ ░███  ░██████   ░░█████████  ░███████████  ░███ ░███      ░███ ███████████ ███████████
//  ░███ ░░░  ░███  ░███░░█    ░░░░░░░░███ ░███░░░░░███  ░███ ░███      ░███░░░░░███░░░ ░░░░░███░░░
//  ░███      ░███  ░███ ░   █ ███    ░███ ░███    ░███  ░███ ░░███     ███     ░███        ░███
//  █████     █████ ██████████░░█████████  █████   █████ █████ ░░░███████░      ░░░         ░░░
// ░░░░░     ░░░░░ ░░░░░░░░░░  ░░░░░░░░░  ░░░░░   ░░░░░ ░░░░░    ░░░░░░░
//
//
//  License:         MIT License
//                   meshio++ default license: LICENSE
//
//  Main authors:    Vicente Mataix Ferrandiz
//
//
// VTK XML MultiBlock. The index carries no geometry at all -- writing means
// carving the mesh into one piece per CellBlock and delegating every piece to
// write_vtu_codec; reading means parsing the index and delegating every piece
// to read_vtu/read_vtp, then combining with operations/merge.hpp.
//
// Anonymous-namespace helpers are prefixed `vtm_`, as the amalgamation
// concatenates every translation unit and `vtu_`/`vts_`/`vtr_` are taken.

// System includes
#include <algorithm>
#include <cstddef>
#include <filesystem>
#include <fstream>
#include <string>
#include <unordered_map>
#include <vector>

// External includes
#include "pugixml.hpp"

// Project includes
#include "meshioplusplus/formats/vtm.hpp"
#include "meshioplusplus/detail/data_ops.hpp"
#include "meshioplusplus/detail/provenance.hpp"
#include "meshioplusplus/exceptions.hpp"
#include "meshioplusplus/formats/vtp.hpp"
#include "meshioplusplus/formats/vtu.hpp"
#include "meshioplusplus/operations/clean.hpp"
#include "meshioplusplus/operations/merge.hpp"
#include "meshioplusplus/region.hpp"
#include "meshioplusplus/detail/classic_stream.hpp"

namespace meshioplusplus {

namespace {

namespace fs = std::filesystem;

// Build a standalone single-block mesh: block `Idx`'s own cells (whichever
// representation -- rectangular, polygon, or polyhedron), the mesh's full
// point array (pruned below), and that block's slice of every cell_data key.
// field_data is not carried onto pieces: it describes the whole mesh, not one
// block, and .vtm has no per-block place to put it (see vtm.md).
Mesh vtm_extract_piece(const Mesh& rMesh, std::size_t Idx) {
    Mesh piece;
    piece.AssignPoints(detail::data_owned_copy(rMesh.Points()));

    const auto cb = rMesh.Cells(Idx);
    if (cb.IsPolyhedron()) {
        std::vector<std::vector<std::vector<std::int64_t>>> cells(cb.NumCells());
        for (std::size_t r = 0; r < cb.NumCells(); ++r) {
            cells[r].resize(cb.NumFaces(r));
            for (std::size_t f = 0; f < cb.NumFaces(r); ++f) {
                const auto face = cb.Face(r, f);
                cells[r][f].assign(face.first, face.first + face.second);
            }
        }
        piece.AddPolyhedronBlock(std::string(cb.Type()), std::move(cells));
    } else if (cb.IsRagged()) {
        std::vector<std::vector<std::int64_t>> rows(cb.NumCells());
        for (std::size_t r = 0; r < cb.NumCells(); ++r)
            rows[r].assign(cb.Row(r), cb.Row(r) + cb.RowSize(r));
        piece.AddPolygonBlock(std::string(cb.Type()), std::move(rows));
    } else {
        piece.AddCellBlock(std::string(cb.Type()), detail::data_owned_copy(cb.Conn()));
    }

    for (const std::string& name : rMesh.PointDataNames())
        piece.AddPointData(name, detail::data_owned_copy(rMesh.PointData(name)));
    for (const std::string& name : rMesh.CellDataNames())
        if (rMesh.CellDataNumBlocks(name) > Idx)
            piece.AddCellData(name, {detail::data_owned_copy(rMesh.CellData(name, Idx))});

    return piece;
}

// Every `<DataSet>` under `pNode`, in document order, regardless of how many
// `<Block>` levels separate them from it (xpath is not linked in, so this is
// a plain recursive walk rather than a `"//DataSet"` select).
void vtm_collect_datasets(const pugi::xml_node& rNode, std::vector<pugi::xml_node>& rOut) {
    for (pugi::xml_node child : rNode.children()) {
        if (std::string(child.name()) == "DataSet")
            rOut.push_back(child);
        else
            vtm_collect_datasets(child, rOut);
    }
}

// One piece, resolved from its `<DataSet>` element: the path to read (relative
// to the index file's own directory) and the name to give it.
struct vtm_piece_ref {
    fs::path mPath;
    std::string mName;
};

std::vector<vtm_piece_ref> vtm_parse_index(const std::string& rPath, pugi::xml_document& rDoc) {
    const pugi::xml_parse_result parsed = rDoc.load_file(rPath.c_str());
    if (!parsed)
        throw ReadError("Could not parse .vtm XML: " + rPath + ": " + parsed.description());

    const pugi::xml_node root = rDoc.child("VTKFile");
    if (!root)
        throw ReadError("Expected tag 'VTKFile': " + rPath);
    if (std::string(root.attribute("type").as_string()) != "vtkMultiBlockDataSet")
        throw ReadError("Expected type vtkMultiBlockDataSet: " + rPath);
    const pugi::xml_node mb = root.child("vtkMultiBlockDataSet");
    if (!mb)
        throw ReadError("Expected tag 'vtkMultiBlockDataSet': " + rPath);

    std::vector<pugi::xml_node> datasets;
    vtm_collect_datasets(mb, datasets);

    const fs::path base = fs::path(rPath).parent_path();
    std::vector<vtm_piece_ref> pieces;
    pieces.reserve(datasets.size());
    for (const pugi::xml_node& ds : datasets) {
        const std::string file = ds.attribute("file").as_string("");
        if (file.empty())
            throw ReadError("<DataSet> is missing its 'file' attribute: " + rPath);
        vtm_piece_ref ref;
        ref.mPath = base.empty() ? fs::path(file) : base / file;
        ref.mName = ds.attribute("name").as_string("");
        if (ref.mName.empty())
            ref.mName = "block_" + std::to_string(pieces.size());
        pieces.push_back(std::move(ref));
    }
    return pieces;
}

}  // namespace

void write_vtm_codec(const std::string& rPath, const Mesh& rMesh, bool binary, detail::VtkCodec codec) {
    if (binary && codec != detail::VtkCodec::None)
        detail::vtk_codec_require_write(codec);

    const fs::path index_path(rPath);
    const std::string stem = index_path.stem().string();
    const fs::path dir = index_path.parent_path().empty() ? fs::path(stem)
                                                            : index_path.parent_path() / stem;
    std::error_code ec;
    fs::create_directories(dir, ec);
    if (ec)
        throw WriteError("Could not create directory for .vtm pieces: " + dir.string() + ": " +
                          ec.message());

    std::vector<std::string> piece_files;
    std::vector<std::string> piece_names;
    piece_files.reserve(rMesh.NumCellBlocks());
    piece_names.reserve(rMesh.NumCellBlocks());

    for (std::size_t i = 0; i < rMesh.NumCellBlocks(); ++i) {
        const Mesh piece = vtm_extract_piece(rMesh, i);
        CleanOptions copts;
        copts.weld = false;
        copts.remove_orphans = true;
        copts.drop_degenerate = false;
        copts.drop_duplicate_cells = false;
        const CleanResult cleaned = clean(piece, copts);

        const std::string piece_file_name = stem + "_" + std::to_string(i) + ".vtu";
        const fs::path piece_path = dir / piece_file_name;
        write_vtu_codec(piece_path.string(), cleaned.mMesh, binary, binary ? codec : detail::VtkCodec::None);

        piece_files.push_back((fs::path(stem) / piece_file_name).generic_string());
        piece_names.push_back("block_" + std::to_string(i));
    }

    auto os = detail::make_classic_ofstream(rPath, std::ios::binary);
    if (!os)
        throw WriteError("Could not open file for writing: " + rPath);

    os << "<?xml version=\"1.0\"?>\n";
    os << "<VTKFile type=\"vtkMultiBlockDataSet\" version=\"1.0\" byte_order=\"LittleEndian\">\n";
    os << detail::provenance_render_xml_comment(detail::SlotTier::Block) << "\n";
    os << "<vtkMultiBlockDataSet>\n";
    os << "<Block index=\"0\">\n";
    for (std::size_t i = 0; i < piece_files.size(); ++i)
        os << "<DataSet index=\"" << i << "\" name=\"" << piece_names[i] << "\" file=\""
           << piece_files[i] << "\"/>\n";
    os << "</Block>\n";
    os << "</vtkMultiBlockDataSet>\n";
    os << "</VTKFile>\n";
    if (!os)
        throw WriteError("Failed while writing: " + rPath);
}

void write_vtm(const std::string& rPath, const Mesh& rMesh, bool binary, bool zlib) {
    write_vtm_codec(rPath, rMesh, binary, zlib ? detail::VtkCodec::Zlib : detail::VtkCodec::None);
}

Mesh read_vtm(const std::string& rPath, const ReadOptions& rOpts) {
    pugi::xml_document doc;
    const std::vector<vtm_piece_ref> refs = vtm_parse_index(rPath, doc);
    if (refs.empty()) {
        Mesh empty;
        empty.AssignPoints(NDArray::Uninit(DType::Float64, {0, 3}));
        return empty;
    }

    std::vector<Mesh> pieces;
    pieces.reserve(refs.size());
    for (const vtm_piece_ref& ref : refs) {
        const std::string ext = ref.mPath.extension().string();
        if (ext == ".vtu")
            pieces.push_back(read_vtu(ref.mPath.string(), rOpts));
        else if (ext == ".vtp")
            pieces.push_back(read_vtp(ref.mPath.string(), rOpts));
        else
            throw ReadError("Unsupported .vtm piece '" + ref.mPath.string() +
                             "': only .vtu/.vtp pieces are read");
    }

    std::vector<const Mesh*> ptrs;
    ptrs.reserve(pieces.size());
    for (const Mesh& m : pieces)
        ptrs.push_back(&m);

    MergeOptions mopts;
    mopts.weld = false;
    mopts.source_tag = true;
    mopts.data_policy = MergeDataPolicy::Fill;
    MergeResult result = merge(ptrs, mopts);

    for (std::size_t i = 0; i < refs.size(); ++i)
        result.mMesh.AddRegion(Region(refs[i].mName, RegionKind::Cell, std::move(result.mCellMaps[i])));

    return std::move(result.mMesh);
}

MeshMetadata read_vtm_metadata(const std::string& rPath, const ReadOptions& rOpts) {
    pugi::xml_document doc;
    const std::vector<vtm_piece_ref> refs = vtm_parse_index(rPath, doc);

    MeshMetadata meta;
    meta.mFormat = "vtm";
    if (refs.empty())
        return meta;

    std::vector<CellBlockInfo> blocks;
    std::unordered_map<std::string, std::size_t> type_to_idx;
    std::vector<std::string> point_names, cell_names, field_names;
    auto merge_names = [](std::vector<std::string>& rInto, const std::vector<std::string>& rFrom) {
        for (const std::string& n : rFrom)
            if (std::find(rInto.begin(), rInto.end(), n) == rInto.end())
                rInto.push_back(n);
    };

    for (const vtm_piece_ref& ref : refs) {
        const std::string ext = ref.mPath.extension().string();
        const MeshMetadata pm = (ext == ".vtp") ? read_vtp_metadata(ref.mPath.string(), rOpts)
                                                 : read_vtu_metadata(ref.mPath.string(), rOpts);
        meta.mNumPoints += pm.mNumPoints;
        meta.mPointDim = std::max(meta.mPointDim, pm.mPointDim);
        meta.mFellBackToFullRead = meta.mFellBackToFullRead || pm.mFellBackToFullRead;
        merge_names(point_names, pm.mPointDataNames);
        merge_names(cell_names, pm.mCellDataNames);
        merge_names(field_names, pm.mFieldDataNames);

        for (const CellBlockInfo& cb : pm.mCellBlocks) {
            auto it = type_to_idx.find(cb.mType);
            if (it == type_to_idx.end()) {
                type_to_idx[cb.mType] = blocks.size();
                blocks.push_back(cb);
            } else {
                CellBlockInfo& dst = blocks[it->second];
                dst.mNumCells += cb.mNumCells;
                dst.mRagged = dst.mRagged || cb.mRagged;
                if (dst.mNodesPerCell != cb.mNodesPerCell)
                    dst.mNodesPerCell = 0;
            }
        }
    }

    meta.mCellBlocks = std::move(blocks);
    std::sort(point_names.begin(), point_names.end());
    std::sort(cell_names.begin(), cell_names.end());
    std::sort(field_names.begin(), field_names.end());
    meta.mPointDataNames = std::move(point_names);
    meta.mCellDataNames = std::move(cell_names);
    meta.mFieldDataNames = std::move(field_names);
    return meta;
}

}  // namespace meshioplusplus

File: src/python/meshioplusplus/vtm/_vtm.py
"""VTK XML MultiBlock, the pure-Python reference (v11.6.0, roadmap §1 tier B4,
part 3 of 3).

The index carries no geometry at all: it is a list of `<DataSet file="..."/>`
entries, each pointing at a standalone `.vtu` (or `.vtp`) piece. Writing means
carving the mesh into one piece per cell block -- pruned to only the points
that block references, via `meshioplusplus.clean` -- and delegating each
piece to the best available `.vtu` writer. Reading means parsing the index,
reading every piece with the best available reader for its own extension,
and combining them with the internal numpy `merge()` (no welding: multiblock
pieces are pre-separated by construction, not coincident-point fragments to
weld back together). Each piece becomes one `"cell"` region in the merged
mesh, named from the index's `name=` attribute, via `merge()`'s own per-input
cell index map -- correct even when same-typed pieces consolidate into one
output cell block.
"""

from __future__ import annotations

import os
import xml.etree.ElementTree as ET

import numpy as np

from .. import _provenance
from .._clean import clean
from .._exceptions import ReadError
from .._merge import _merge_py
from .._mesh import Mesh
from .._pvtk_index import share_field_data
from .._regions import Region


def _collect_datasets(node, out):
    for child in node:
        if child.tag == "DataSet":
            out.append(child)
        else:
            _collect_datasets(child, out)


def _read_piece(path):
    ext = os.path.splitext(path)[1]
    if ext == ".vtu":
        from .. import vtu as pkg
    elif ext == ".vtp":
        from .. import vtp as pkg
    else:
        raise ReadError(
            f"Unsupported .vtm piece '{path}': only .vtu/.vtp pieces are read"
        )
    return pkg.read(path)


def read(filename):
    tree = ET.parse(str(filename))
    root = tree.getroot()
    if root.tag != "VTKFile":
        raise ReadError("Expected tag 'VTKFile'")
    if root.get("type") != "vtkMultiBlockDataSet":
        raise ReadError("Expected type vtkMultiBlockDataSet")
    mb = root.find("vtkMultiBlockDataSet")
    if mb is None:
        raise ReadError("Expected tag 'vtkMultiBlockDataSet'")

    datasets = []
    _collect_datasets(mb, datasets)
    if not datasets:
        return Mesh(np.zeros((0, 3), dtype=np.float64), [])

    base = os.path.dirname(str(filename))
    pieces = []
    names = []
    for i, ds in enumerate(datasets):
        file_attr = ds.get("file")
        if not file_attr:
            raise ReadError("<DataSet> is missing its 'file' attribute")
        path = os.path.join(base, file_attr) if base else file_attr
        pieces.append(_read_piece(path))
        names.append(ds.get("name") or f"block_{i}")

    out, _point_maps, cell_maps = _merge_py(
        pieces,
        weld=False,
        atol=1e-8,
        source_tag=True,
        data_policy="fill",
        drop_duplicate_cells=False,
    )
    # Dataset-global field data is the union across pieces, not namespaced.
    share_field_data(out, pieces)
    for name, cmap in zip(names, cell_maps):
        out.regions.append(Region(name, "cell", np.asarray(cmap, dtype=np.int64)))
    return out


def write(filename, mesh, binary=True, compression="zlib", header_type=None):
    filename = str(filename)
    stem = os.path.splitext(os.path.basename(filename))[0]
    parent = os.path.dirname(filename)
    piece_dir = os.path.join(parent, stem) if parent else stem
    os.makedirs(piece_dir, exist_ok=True)

    from .. import vtu as vtu_pkg

    piece_files = []
    piece_names = []
    for i, cb in enumerate(mesh.cells):
        piece = Mesh(
            mesh.points,
            [(cb.type, cb.data)],
            point_data=dict(mesh.point_data),
            cell_data={k: [v[i]] for k, v in mesh.cell_data.items() if i < len(v)},
        )
        cleaned = clean(
            piece,
            weld=False,
            remove_orphans=True,
            drop_degenerate=False,
            drop_duplicate_cells=False,
        )
        piece_file_name = f"{stem}_{i}.vtu"
        vtu_pkg.write(
            os.path.join(piece_dir, piece_file_name),
            cleaned,
            binary=binary,
            compression=compression,
            header_type=header_type,
        )
        piece_files.append(f"{stem}/{piece_file_name}")
        piece_names.append(f"block_{i}")

    lines = ['<?xml version="1.0"?>']
    lines.append(
        '<VTKFile type="vtkMultiBlockDataSet" version="1.0" byte_order="LittleEndian">'
    )
    lines.append(_provenance.render_xml_comment(_provenance.SlotTier.BLOCK))
    lines.append("<vtkMultiBlockDataSet>")
    lines.append('<Block index="0">')
    for i, (f, name) in enumerate(zip(piece_files, piece_names)):
        lines.append(f'<DataSet index="{i}" name="{name}" file="{f}"/>')
    lines.append("</Block>")
    lines.append("</vtkMultiBlockDataSet>")
    lines.append("</VTKFile>")
    with open(filename, "w") as f:
        f.write("\n".join(lines) + "\n")
